### DISCUSSION TITLE
Google OAuth error "400: invalid request"

### DISCUSSION THREAD

**Discussion by dfluff** [NestMTX/app#35] [Q&A]
Hi,
I'm hitting a brick wall when I try to authenticate my credentials.  I've got the OAuth client id, secret, and Device Access Project ID (project has the same OAuth client ID) all set up, but when I try to authenticate I get an "Access blocked: authorisation error" from the google side with a "Error 400: invalid_request".
The one questionable part that I can think of is the redirect URI I set for the OAuth config.  Google wouldn't let me use https://nas:2001 or even nas.local, it has to be a valid tld, so I used a URI that is valid and resolvable and that, _from within my network_ will resolve back to the NestMTX docker container.  I think that should be valid but might not be, and is the only thing I can think of that might be causing an issue.
I managed to go through a similar process to set up the Nest integration in Home Assistant yesterday and the OAuth worked fine for that, so I'm at a bit of a loss as to what might be different here.

**Comment by john-jaffe** [comment]
I ran into a similar issue. Accessing the nestmtx console through "http://server.domainlocal.net:2000" instead of the shortname seemed to have fixed it for me. That way the redirect URI that it showed to copy from the credentials page matched what I put in the authorized redirect URI in the google oauth console. Either that, or it was a coincidence and it was just taking a while for google to apply the new URI.

**Reply by jakguru** [reply]
I've heard anecdotal reports that I haven't tested myself that Google requires a proper top level domain for OAuth redirects. 

But Google does seem to take a longgggg time to update its redirects. So much so that it's documented in our docs:

https://nestmtx.com/setup/credentials.html#authorizing-credentials


> Getting errors about invalid redirect URL even though you set the right one? That's a common issue caused by a long cacheing interval on Google's end. You may need to wait up to an hour for an update to the credentials to be applied by Google but usually it's closer to about 5 minutes.

**Comment by dfluff** [comment]
Hmm.. I've given it several days with no change.  The error gives two links - one to "find out more" about the error which just redirects to my Google account page and the other for docs that goes to https://developers.google.com/identity/protocols/oauth2/policies#secure-response-handling ...suggesting that it's not happy about non-secure origins and redirects, BUT the page in NestMTX I'm coming from is https and the redirect URI is also https, points back to me with a valid public DNS, valid TLS cert, etc.
I'll try some more digging later, but I'm not hopeful without some more useful info coming out of the Google side of things.

**Reply by jakguru** [reply]
Please feel free to join our discord server: https://discord.gg/hMAEuNa4Fd

From there you can ping me 1:1 and we can go through some additional troubleshooting. A screen recording would be useful, but I don't want to ask you to publish that publically.

**Comment by dfluff** [comment]
Thanks for the offer of help @jakguru, that's very generous!
I appear to have just managed to get OAuth working.  Annoyingly I'm not completely sure what I changed... I removed and re-set the URI in Google (although I had done that previously), double-checked that the FQDN I was giving it was valid on the internet, etc.. and for some reason this time it worked.

..although no camera appeared in NestMTX.  I think my earlier disparate attempts at set-up had got wires crossed because if I hit the Edit icon for the now working Credentials the Google OAuth page showed me the name of a _different_ one of my GoogleCloud project (or possibly Device Access projects - I have re-used the names making it hard to know which it's referring to).  I've double checked all the IDs but kept getting the same result.

...SO what I have eventually done is ditch the credentials (and GCP project) I made for NestMTX, used the GCP I set up for Nest access in Home Assistant (by adding the NestMTX redirect URI as a second URI to that project) and bingo!  The credentials have authenticated and the camera is showing up (awaiting configuration) in the Cameras list.

As I say, I cant work out why my earlier attempts didn't work.  I can only assume that following an initial mis-configuration I then made matters worse with my later fault-finding.

Thanks again for your helpful responses.

**Reply by jakguru** [reply]
If you were following the tutorial for how to setup the GCP & DAC, there might have been some information which is out of date. If that is the case, I apologize for the inconvenience and if you're able to pinpoint what went wrong I will be happy to update the docs to reflect that.
